### PR TITLE
[MHLO]: variadic length input and different types of reduce op lower …

### PR DESCRIPTION
**Pull request by lipracer**
…to linalg

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 1]
```diff
@@ -1952,23 +1952,23 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
...
-    auto src_type = adaptor.inputs()[0].getType().cast<ShapedType>();
-    int src_rank = src_type.getRank();
-    if (!src_rank) {
+    auto ranked_iter = std::find_if(
+        std::begin(adaptor.inputs()), std::end(adaptor.inputs()),
+        [](Value v) { return v.getType().cast<ShapedType>().getRank(); });
+
+    if (ranked_iter == std::end(adaptor.inputs())) {
       return rewriter.notifyMatchFailure(op, "expects known-rank args");
     }
```
[the remark this answers is not in the document]
Comment (lipracer) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1961]: all of reduce's inputs has same rank, I think just need to find the ranked tensor.

**Review by stellaraccident — APPROVED** [review]
Thank you so much - this was a pleasant surprise to find that someone had fixed https://github.com/google/iree/issues/7748

**Review by jpienaar — APPROVED** [review]

**Review by joker-eph — APPROVED** [review]
Approving again, there seems to be an issue with the infra, if it does not go through we may need to rebase this (or I'll import manually)

**Comment by lipracer** [comment]
I have rebased on the latest.

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 2]
```diff
@@ -1952,23 +1952,23 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
...
 
     SmallVector<int64_t, 4> reduction_dims = Extract1DVector(op.dimensions());
 
     SmallVector<Value> inputs, outputs;
     SmallVector<AffineMap, 3> indexing_maps;
-    for (int i = 0; i < num_inputs; ++i) {
-      Value src = adaptor.inputs()[i];
-      if (src.getType() != src_type) return failure();
-
+    for (auto it : llvm::enumerate(adaptor.inputs())) {
```
[the remark this answers is not in the document]
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1557]: You marked it as resolved but the code isn't updated?

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc [hunk 3]
```diff
@@ -1611,8 +1609,12 @@ class ReduceOnTensorsConversion : public OpConversionPattern<mhlo::ReduceOp> {
     Region& region = linalg_op.region();
     rewriter.inlineRegionBefore(op.body(), region, region.end());
     TypeConverter::SignatureConversion signature_converter(num_inputs * 2);
-    for (int i = 0; i < num_inputs * 2; ++i)
-      signature_converter.addInputs(i, src_type.getElementType());
+
+    // map input and init values's types
+    for (auto it : llvm::enumerate(op.getOperation()->getOperands()))
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc:1614]: clang-tidy complains about this (and annoyingly this is blocking the commit), it wants `const auto &it` here.

**Review by joker-eph — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (joker-eph) [reply tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc]: Can you comment on this? You updated the PR again but you didn't answer here, nor applied the change I believe?

**Comment by joker-eph** [comment]
See the conversion inline above : https://github.com/tensorflow/tensorflow/pull/53232#discussion_r758829334

**Comment by lipracer** [comment]
> See the conversion inline above : [#53232 (comment)](https://github.com/tensorflow/tensorflow/pull/53232#discussion_r758829334)

Sorry. I misread it.

**Comment by lipracer** [comment]
> See the conversion inline above : [#53232 (comment)](https://github.com/tensorflow/tensorflow/pull/53232#discussion_r758829334)

@joker-eph  Sorry. I misread it. please review again.

**Review by rthadur — APPROVED** [review]

**Comment by lipracer** [comment]
Rebase on latest.

**Review by rthadur — APPROVED** [review]

**Comment by lipracer** [comment]
windows bazel build error log:
 ```external/llvm-project/llvm/lib/IR/AutoUpgrade.cpp(3726): fatal error C1061: compiler limit: blocks nested too deeply```
It seems that visual studio does not support variadic templates very well。：）
Should I update the code base? I think it's not my PR problem.

**Comment by lipracer** [comment]
I found the llvm build in the error log.I think it’s not caused by my code. I don't have a windows environment, should I build one or update my code base?

**Review by rthadur — APPROVED** [review]

**Pull request merged**